### PR TITLE
[BPK-2596] Add Transit Information API to Tappable Link Labels

### PR DESCRIPTION
**Pull request by georgegillams**
+ [x] Check this if you have read and followed the [contributing guidelines](https://github.com/Skyscanner/backpack-ios/blob/master/CONTRIBUTING.md)

+ [x] Readme

### Still todo:
 - [x] Update delegate method API to say `tappableLabel` instead of `attributedLabel`
 - [ ] Fix issue where text is cut off

### CODE REVIEW

File: Backpack/TappableLinkLabel/README.md [hunk 1]
```diff
@@ -11,20 +11,27 @@
 
 BPKTappableLinkLabel *tappableLinkLabel = [[BPKTappableLinkLabel alloc] initWithFontStyle:BPKFontStyleTextBase];
 tappableLinkLabel.text = @“Find out more about our terms of service.”
-tappableLinkLabel.delegate = self;
+[tappableLinkLabel setDelegate:self];
```
[the remark this answers is not in the document]
Comment (georgegillams) [reply Backpack/TappableLinkLabel/README.md]: That was an accidental regression...

File: Backpack/TextView/Classes/BPKTextView.m [hunk 2]
```diff
@@ -80,43 +80,25 @@ - (void)setText:(NSString *_Nullable)text {
...
-
-    [self updateStyle];
-}
-
 - (void)setFontStyle:(BPKFontStyle)fontStyle {
     BPKAssertMainThread();
     _fontStyle = fontStyle;
 
-    [self updateStyle];
+    [self setText:self.attributedText.string];
```
[the remark this answers is not in the document]
Comment (georgegillams) [reply Backpack/TextView/Classes/BPKTextView.m]: This is an unrelated change that I've now reverted

File: Example/Backpack/View Controllers/TappableLinkLabelsSelectorViewController.swift [hunk 3]
```diff
@@ -33,6 +33,10 @@ class TappableLinkLabelsSelectorViewController: UITableViewController {
             segue.destination.title = "Alternate style links"
             maybeTappableLinkLabelsController?.linkCount = .double
             maybeTappableLinkLabelsController?.style = .alternate
+        case "NonURLLinks":
```
[the remark this answers is not in the document]
Comment (georgegillams) [reply Example/Backpack/View Controllers/TappableLinkLabelsSelectorViewController.swift]: I've made the change, but tbh I think for the way we use them this isn't actually an improvement

File: Backpack/TappableLinkLabel/README.md [hunk 4]
```diff
@@ -14,17 +14,24 @@ tappableLinkLabel.text = @“Find out more about our terms of service.”
...
 
 ...
 
-#pragma BPKTappableLabelDelegate
+#pragma mark - <BPKTappableLinkLabelDelegate>
 
-- (void)attributedLabel:(BPKTappableLinkLabel *)label didSelectLinkWithURL:(NSURL *)url {
-    [delegate openURL:url];
+- (void)attributedLabel:(BPKTappableLinkLabel *)label didSelectLinkWithURL:(NSURL *)url
+{
```
Comment (k0nserv) [inline Backpack/TappableLinkLabel/README.md]: This doens't match our brace style

File: Backpack/TappableLinkLabel/README.md [hunk 5]
```diff
@@ -14,17 +14,24 @@ tappableLinkLabel.text = @“Find out more about our terms of service.”
...
+    [_presenter urlTapped:url];
+}
+
+- (void)attributedLabel:(BPKTappableLinkLabel *)label didSelectLinkWithTransitInformation:(NSDictionary *)components
+{
```
Comment (k0nserv) [inline Backpack/TappableLinkLabel/README.md]: Same as above

File: Example/Backpack/View Controllers/TappableLinkLabelsSelectorViewController.swift (hunk 3 above)
[the remark this answers is not in the document]
Comment (k0nserv) [reply Example/Backpack/View Controllers/TappableLinkLabelsSelectorViewController.swift]: Yeah it's not much better bacause we are setting the values in Storyboard.

File: Backpack/TappableLinkLabel/Classes/BPKTappableLinkLabel.m [hunk 6]
```diff
@@ -160,16 +163,22 @@ - (void)updateTextDisplay {
...
     self.contentView.text = newString;
 
     // Re-apply the links
     for (BPKTappableLinkDefinition *linkDefinition in _persistedLinks) {
-        [self.contentView addLinkToURL:linkDefinition.url withRange:linkDefinition.range];
+        if (linkDefinition.hasURLDefinition) {
+            [self.contentView addLinkToURL:linkDefinition.url withRange:linkDefinition.range];
+        } else if (linkDefinition.hasTransitInformationDefinition) {
+            [self.contentView addLinkToTransitInformation:linkDefinition.components withRange:linkDefinition.range];
+        }
```
Comment (k0nserv) [inline Backpack/TappableLinkLabel/Classes/BPKTappableLinkLabel.m]: Might be good to have an assert here in `else`

File: Backpack/TappableLinkLabel/Classes/BPKTappableLinkLabel.m [hunk 7]
```diff
@@ -240,11 +256,17 @@ - (NSInteger)numberOfLines {
...
 - (void)attributedLabel:(BPKTappableLinkLabel *)label didSelectLinkWithURL:(NSURL *)url {
-    if (self.delegate != nil) {
+    if ([self.delegate respondsToSelector:@selector(attributedLabel:didSelectLinkWithURL:)]) {
         [self.delegate attributedLabel:label didSelectLinkWithURL:url];
     }
 }
 
+- (void)attributedLabel:(BPKTappableLinkLabel *)label didSelectLinkWithTransitInformation:(NSDictionary *)components {
+        if ([self.delegate respondsToSelector:@selector(attributedLabel:didSelectLinkWithTransitInformation:)]) {
+        [self.delegate attributedLabel:label didSelectLinkWithTransitInformation:components];
```
Comment (k0nserv) [inline Backpack/TappableLinkLabel/Classes/BPKTappableLinkLabel.m]: The indentation is messed up here

File: Backpack/TappableLinkLabel/Classes/BPKTappableLinkLabel.m (hunk 6 above)
Comment (georgegillams) [reply Backpack/TappableLinkLabel/Classes/BPKTappableLinkLabel.m]: done

File: Backpack/TappableLinkLabel/Classes/BPKTappableLinkLabel.m (hunk 7 above)
Comment (georgegillams) [reply Backpack/TappableLinkLabel/Classes/BPKTappableLinkLabel.m]: done

**Pull request merged**